Launch: Cleanup calypsoifyGutenberg.

yansern opened the pull request:
#### Changes proposed in this Pull Request

Note: Ensure ETK is deployed with changes from https://github.com/Automattic/wp-calypso/pull/49004 before deploying this one.

This PR branches off https://github.com/Automattic/wp-calypso/pull/49004 to cleanup props that are no longer used in `calypsoifyGutenberg`.

The reason for this PR's existence is to keep deploys atomic. https://github.com/Automattic/wp-calypso/pull/49004 is related to ETK and this is related to WBE.

~~TODO: The `currentCalypsoUrl` is the only thing left that keeps Focused Launch tethered to Calypso I think. Need to see how to move that out.~~ This is needed for for "Use A Domain I Own" to identify the correct url on the parent iframe.

These feature flags are also removed:
- gutenboarding/new-launch-mobile
- gutenboarding/feature-picker
- create/persistent-launch-button
- create/focused-launch-flow

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Ensure Step-By-Step Launch & Focused Launch still works.
* Also good to check who is accessing props from `calypsoifyGutenberg` object, most places are still relying on `isGutenboarding` and that is why it is still kept.

Fixes https://github.com/Automattic/wp-calypso/issues/46839 #47803 #50089

yansern committed:
Experimental codes for focused launch a/b test.

yansern committed:
Clean up unused prop on calypsoifyGutenberg object.

yansern commented:
Ignore D55672-code, this was generated because the PR has an initial incorrect base branch.

First deploy the WBE, then deploy Calypso.

yansern committed:
Revert "Experimental codes for focused launch a/b test."

This reverts commit 614660e8902b1f06ceeedfa6d7129c0126775562.

yansern commented:
Rebased and this commit https://github.com/Automattic/wp-calypso/pull/49067/commits/3a3d2a6f67aa8cbe026f90d950ab251e710e4f12 was reverted because it was a rebase glitch.

ciampo commented:
Hey @yansern , do you mind rebasing this branch, since it's become slightly stale (6/7 days ago?)

ciampo committed:
Remove duplicate window types, move correct types to ETK package

ciampo commented:
Waiting until Focused Launch is released

ciampo reviewed (approved):
LGTM

ciampo committed:
Add manageReusableBlocksUrl to calypsoifyGutenberg types

razvanpapadopol commented on apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js:
```diff
@@ -790,23 +790,8 @@ function getGutenboardingStatus( calypsoPort ) {
-			isNewLaunchMobile,
-			isExperimental,
-			isPersistentLaunchButton,
-			isFocusedLaunchFlow,
```
Let's not remove this yet since it's still in use: https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts#L71

ciampo replied:
Resolving the conversation, since the use of the `isFocusedLaunchFlow` is being removed with this same PR now

razvanpapadopol commented on client/gutenberg/editor/calypsoify-iframe.tsx:
```diff
@@ -376,22 +376,10 @@ class CalypsoifyIframe extends Component<
-			const isNewLaunchMobile = config.isEnabled( 'gutenboarding/new-launch-mobile' ); // TODO: remove after ETK 2.8.6 is released
-			const isExperimental = config.isEnabled( 'gutenboarding/feature-picker' ); // TODO: remove after ETK 2.8.6 is released
-			const isPersistentLaunchButton = config.isEnabled( 'create/persistent-launch-button' ); // TODO: remove after ETK 2.8.6 is released
-			const isFocusedLaunchFlow = config.isEnabled( 'create/focused-launch-flow' );
```
Same as above: https://github.com/Automattic/wp-calypso/pull/49067#discussion_r570304869

yansern replied:
Okay so these 3 feature flags are now removed.

razvanpapadopol commented on client/gutenberg/editor/calypsoify-iframe.tsx:
```diff
@@ -376,22 +376,10 @@ class CalypsoifyIframe extends Component<
 			const isGutenboarding = this.props.siteCreationFlow === 'gutenboarding';
-			const isSiteUnlaunched = this.props.isSiteUnlaunched;
-			const launchUrl = `${ window.location.origin }/start/launch-site?siteSlug=${ this.props.siteSlug }`;
-			const isNewLaunchMobile = config.isEnabled( 'gutenboarding/new-launch-mobile' ); // TODO: remove after ETK 2.8.6 is released
-			const isExperimental = config.isEnabled( 'gutenboarding/feature-picker' ); // TODO: remove after ETK 2.8.6 is released
-			const isPersistentLaunchButton = config.isEnabled( 'create/persistent-launch-button' ); // TODO: remove after ETK 2.8.6 is released
```
Shouldn't we also cleanup the feature flags from config files?

yansern replied:
Okay so these 3 feature flags are now removed.

yansern committed:
Remove focused launch flow and persistent button feature flag.

yansern committed:
Removed feature picker and new launch mobile feature flags.

yansern commented:
Removed the feature flags:
- gutenboarding/new-launch-mobile
- gutenboarding/feature-picker
- create/persistent-launch-button
- create/focused-launch-flow

And also codes related to `isFocusedLaunchFlowFeatureFlag` for temp feature flag.

Also rebased the PR.

StefanNieuwenhuis commented:
> Removed the feature flags:
> 
> * gutenboarding/new-launch-mobile
> * gutenboarding/feature-picker
> * create/persistent-launch-button
> * create/focused-launch-flow
> 
> And also codes related to `isFocusedLaunchFlowFeatureFlag` for temp feature flag.
> 
> Also rebased the PR.

I've created #50089 to document this.

ciampo commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts:
```diff
@@ -67,14 +67,11 @@ domReady( () => {
 			}
 
-			if ( launchFlow === SITE_LAUNCH_FLOW && ! isFocusedLaunchFlowFeatureFlag ) {
+			if ( launchFlow === SITE_LAUNCH_FLOW ) {
```
I would either:
- add a `return`  clause inside all these ifs blocks, or
- re-write the ifs as a `if (..) { … } else if (…) else { … }` chain, or
- use a `switch` statement instead for all the `launchFlow` checks

Basically, this would be to make it explicit that only one type of launch can happen at the same time.

Bonus question: what does the `dispatch( ‘automattic/launch’ ).enableAnchorFm()` call do? If that line is called, should we still run through the rest of the file, or can we _early `return`_?

yansern replied:
> add a `return` clause inside all these ifs blocks,

Possible with some wrangling... there is a part at the end which calls `savePost()`. This happens in the background for gutenboardingLaunch and focusedLaunch, whereas, this needs to be happen first before siteLaunch.

We'll kinda end up with repeated `savePost()` but I think the readability will improve.

> use a `switch` statement instead for all the launchFlow checks

Possible too, but I think I'm in favour of separate `if` blocks. Not have to add `break;` is nice.


> Bonus question: what does the dispatch( ‘automattic/launch’ ).enableAnchorFm() call do? If that line is called, should we still run through the rest of the file, or can we early return?

Anchor-flavoured sites skip some steps in the gutenboarding launch flow (and they are exclusively used the gutenboarding launch flow). `enableAnchorFm` toggles a flag in the data store that changes the step sequence, then the step-by-step launch opens.

ciampo replied:
I see, thank you for the explanation

> there is a part at the end which calls savePost(). This happens in the background for gutenboardingLaunch and focusedLaunch, whereas, this needs to be happen first before siteLaunch.

I missed that part! In that case, we can probably leave it as is — I'm not sure that any refactor would improve the code clarity much

yansern replied:
I'm going to leave the `switch` version here for reference.
I'm not sure why there was a `setTimeout` on the `savePost()` for gutenboarding launch & focused launch.
It's either because we don't want user to see the post is saving or we don't want 2 things to happen at once (slow down the launch popup?).
This switch version reuses one `savePost()` for all without the `setTimeout`.

```js
const saveTask = dispatch( 'core/editor' ).savePost();

switch ( launchFlow ) {
	case GUTENBOARDING_LAUNCH_FLOW:
		dispatch( 'automattic/launch' ).openSidebar();
		break;

	case FOCUSED_LAUNCH_FLOW:
		dispatch( 'automattic/launch' ).openFocusedLaunch();
		break;

	case SITE_LAUNCH_FLOW:
		// Save post first before redirecting to launch url
		( async () => {
			await saveTask;
			window.top.location.href = launchUrl;
		} )();
		break;
}
```

ciampo replied:
I think that the delay was there to:
- prioritise any API call that the launch modal would make
- avoid showing the "post saved" toast in the bottom left of the screen (by adding a delay, the toast would appear while the launch modal is already visible, and therefore the user would not see it)

This is one possible way of using a `switch` statement while keeping the `setTimeout` call

```typescript
const { savePost } = dispatch( 'core/editor' );
const delayedSavePost = () => setTimeout( () => savePost(), 1000 );

switch ( launchFlow ) {
	case GUTENBOARDING_LAUNCH_FLOW:
		// Save post in the background while step-by-step flow opens
		dispatch( 'automattic/launch' ).openSidebar();
		delayedSavePost();
		break;
	case FOCUSED_LAUNCH_FLOW:
		// Save post in the background while focused launch flow opens
		dispatch( 'automattic/launch' ).openFocusedLaunch();
		delayedSavePost();
		break;
	case SITE_LAUNCH_FLOW:
        // Save post first before redirecting to launch url
		( async () => {
			await savePost();
			window.top.location.href = launchUrl;
		} )();
		break;
}
```

ciampo replied:
Actually, I managed to track down the PR that introduced the 1s delay #44909 — @razvanpapadopol , is there a reason why we introduced that arbitrary delay?

yansern replied:
Thanks for the `switch` version which seems like a good middle ground for now. :) Delaying save post does have its benefit regardless of the reason, and this block of code is readable, much readable than the existing one where it's executed at the bottom. I'm committing this in.

razvanpapadopol replied:
I don't remember exactly why it's 1s delay. Maybe something related to how long it would take for the launch modal to open? Eg: with a slide-in animation

ciampo replied:
Alright — let's keep the 1s delay in, for now

ciampo commented on apps/editing-toolkit/typings/fse/index.d.ts:
```diff
@@ -1,3 +1,22 @@
 declare module 'a8c-fse-common-data-stores' {}
+
+interface Window {
+	wpcomEditorSiteLaunch?: {
+		launchUrl: string;
+		launchFlow: string;
+		// property does not exist when not isGutenboarding
+		// property holds the value '1' when isGutenboarding
+		isGutenboarding?: '1';
+		locale?: string;
+	};
```
Are we missing `anchorFmPodcastId` ? From a quick look, it seems like it could be either `string` or `undefined`

yansern replied:
Thanks! Will add it in.

yansern committed:
Add anchorFmPodcastId to typings.

yansern committed:
Set a non launchUrl value to the launch button's href attr.

yansern commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts:
```diff
@@ -44,7 +44,11 @@ domReady( () => {
 		// Wrap 'Launch' button link to control launch flow.
 		const launchButton = document.createElement( 'a' );
-		launchButton.href = launchUrl;
+		// This was previously launchUrl but setting it to javascript: void(0)
+		// to prevent users from doing through site launch when they are
+		// assigned to focused launch. This can happen if they hit CMD+Click
+		// or Right Click + Open In New Tab.
+		launchButton.href = 'javascript: void(0);';
```
Though unlikely, this *might* be one reason users assigned to focused launch still end up with site launch.

yansern commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts:
```diff
@@ -44,7 +44,11 @@ domReady( () => {
 		// Wrap 'Launch' button link to control launch flow.
 		const launchButton = document.createElement( 'a' );
-		launchButton.href = launchUrl;
+		// This was previously launchUrl but setting it to javascript: void(0)
+		// to prevent users from doing through site launch when they are
+		// assigned to focused launch. This can happen if they hit CMD+Click
+		// or Right Click + Open In New Tab.
+		launchButton.href = 'javascript: void(0);';
```
Explanation for this change: Though unlikely, this **might** be the reason users assigned to focused launch still end up with site launch.

ciampo replied:
Why don't we make it a `<button>` instead? It would make it more semantically correct and avoid the same issues (plus, this button is already injected via JS)

yansern replied:
Good point. Changing it to a button.

ciampo replied:
Awesome, thank you!

yansern committed:
Use button instead.

yansern committed:
Use switch.

ciampo commented:
Rebased on top of trunk

ciampo committed:
Fix getCurrentLaunchFlowUrl logic and update types

ciampo reviewed (approved):
Overall, the code looks much cleaner! Awesome job @yansern 


I gave it another round of testing after recent changes:

- [x] Step by Step launch modal opens correctly
- [x] Focused Launch modal opens correctly when A/B testing assigned to "treatment"
- [x] Focused Launch modal is not affected by the `create/focused-launch-flow` feature flag
- [x] Control Launch flow happens when Focused Launch A/B testing assigned to "control"


I guess the order of operations is going to be:

1. ~Deploy WBE~ Actually not necessary, as the changes to be deploy are just deletions of deprecated properties 
2. Merge Calypso
3. Release ETK

ciampo commented:
Let's wait until #48452 is merged, rebase on top of trunk, and see if there's any changes needed (e.g. related to `getCurrentLaunchFlowUrl`)

yansern commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/utils.ts:
```diff
@@ -5,24 +5,8 @@ import { doAction, hasAction } from '@wordpress/hooks';
-	}
-};
+export const getCurrentLaunchFlowUrl = (): string =>
+	window?.calypsoifyGutenberg?.currentCalypsoUrl ?? window.location.href;
```
Swithed over to `??` version after rebase. cc @ciampo

yansern committed:
Fix rebase typing errors.

yansern committed:
Temporarily use both currentSiteId and _currentSiteId.

yansern committed:
No need to check isInIframe because getCurrentLaunchFlowUrl() does it already.

ciampo commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/attach-focused-launch.tsx:
```diff
@@ -44,10 +44,7 @@ registerPlugin( 'a8c-editor-editor-focused-launch', {
-			inIframe() ? getCurrentLaunchFlowUrl() : window.location.href,
-			IMMEDIATE_LAUNCH_QUERY_ARG
-		);
+		const shouldLaunch = hasQueryArg( getCurrentLaunchFlowUrl(), IMMEDIATE_LAUNCH_QUERY_ARG );
```
The `inIframe()` check was redundant, since `getCurrentLaunchFlowUrl()` already uses `window.location.href` when in Wp Admin
